### PR TITLE
Prevent syncer to cancel all other services

### LINKED ISSUE #1398: cancel_token.exceptions.OperationCancelled: Cancellation requested by BeaconChainSyncer:Node token
* Trinity version: 123e171a3b673f7d5d5372c71f808b5145ae52f8
* OS: osx
* Python version (output of `python --version`): Python 3.7.5
* Environment (output of `pip freeze`):
**libp2p==0.1.4**

### What is wrong?

Command
```
python eth2/beacon/scripts/run_beacon_nodes.py
```

Error messages
```
alice@16Uiu	    INFO  2019-12-15 23:47:41,026      BCCReceiveServer  Timed out waiting for <trinity.protocol.bcc_libp2p.servers.BCCReceiveServer object at 0x10a690310> to finish its cleanup, forcibly cancelling pending tasks and exiting anyway
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,027      BCCReceiveServer  <trinity.protocol.bcc_libp2p.servers.BCCReceiveServer object at 0x10a690310>: Pending tasks (3): [<Task pending coro=<BCCReceiveServer._handle_beacon_attestation_loop() running at /Users/hwwang/ethereum/trinity/p2p/service.py:158> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10a6cd690>()]> cb=[gather.<locals>._done_callback() at /Users/hwwang/.pyenv/versions/3.7.5/lib/python3.7/asyncio/tasks.py:691]>, <Task pending coro=<BCCReceiveServer._handle_beacon_block_loop() running at /Users/hwwang/ethereum/trinity/p2p/service.py:158> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10a7beb50>()]> cb=[gather.<locals>._done_callback() at /Users/hwwang/.pyenv/versions/3.7.5/lib/python3.7/asyncio/tasks.py:691]>, <Task finished coro=<BCCReceiveServer._process_orphan_blocks_loop() done, defined at /Users/hwwang/ethereum/trinity/p2p/service.py:154> result=None>]
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,027      BCCReceiveServer  All tasks finished
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,027      BCCReceiveServer  <trinity.protocol.bcc_libp2p.servers.BCCReceiveServer object at 0x10a690310> halted cleanly
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040                  Node  Cancelling service: <trinity.protocol.bcc_libp2p.node.Node object at 0x10a5d4a90>
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040                  Node  <trinity.protocol.bcc_libp2p.node.Node object at 0x10a5d4a90> finished: Cancellation requested by Node token
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040                  Node  <trinity.protocol.bcc_libp2p.node.Node object at 0x10a5d4a90> halted cleanly
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040                  Node  <trinity.protocol.bcc_libp2p.node.Node object at 0x10a5d4a90> finished cleanly
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040             DBManager  <trinity.db.manager.DBManager object at 0x10e22bfd0>: closing client connection: Connection closed
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,040  AsyncioEventBusService  Cancelling service: <trinity.extensibility.event_bus.AsyncioEventBusService object at 0x1056807d0>
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,041              Endpoint  Endpoint[bbeacon-node]: server stopped
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,042            Connection  Failed to send SubscriptionsUpdated(subscriptions=set(), response_expected=False). Handler closed.
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,042              Endpoint  Endpoint[bbeacon-node]: stopped
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,043  AsyncioEventBusService  <trinity.extensibility.event_bus.AsyncioEventBusService object at 0x1056807d0> finished: Cancellation requested by AsyncioEventBusService token
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,043  AsyncioEventBusService  <trinity.extensibility.event_bus.AsyncioEventBusService object at 0x1056807d0> halted cleanly
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,043  AsyncioEventBusService  <trinity.extensibility.event_bus.AsyncioEventBusService object at 0x1056807d0> finished cleanly
alice@16Uiu	   DEBUG  2019-12-15 23:47:41,043           IPCListener  <trinity._utils.logging.IPCListener object at 0x10fda42d0>: closing client connection: Connection closed
bob@16Uiu	   DEBUG  2019-12-15 23:47:41,044                 mplex  mplex unavailable while waiting for incoming: failed to read messages correctly from the underlying connection
bob@16Uiu	   DEBUG  2019-12-15 23:47:41,045                pubsub  fail to read from peer 16Uiu2HAmRkNuKCY8UVcYRaaoqRD71e841aEJV9DVXnHBZs34Jop2, error=,closing the stream and remove the peer from record
bob@16Uiu	   DEBUG  2019-12-15 23:47:41,045             gossipsub  removing peer 16Uiu2HAmRkNuKCY8UVcYRaaoqRD71e841aEJV9DVXnHBZs34Jop2
bob@16Uiu	   DEBUG  2019-12-15 23:47:41,045                pubsub  removed dead peer 16Uiu2HAmRkNuKCY8UVcYRaaoqRD71e841aEJV9DVXnHBZs34Jop2
alice@16Uiu	/Users/hwwang/.pyenv/versions/3.7.5/envs/env37a/lib/python3.7/site-packages/eth_keys/datatypes.py:28: DeprecationWarning: The eth_utils.typing module will be deprecated in favor of eth-typing in the next major version bump.
alice@16Uiu	  from eth_utils.typing import (
alice@16Uiu	/Users/hwwang/.pyenv/versions/3.7.5/envs/env37a/lib/python3.7/site-packages/ssz/sedes/serializable.py:97: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
alice@16Uiu	  class BaseSerializable(collections.Sequence):
alice@16Uiu	/Users/hwwang/ethereum/trinity/trinity/extensibility/asyncio.py:60: ResourceWarning: unclosed resource <TCPServer closed=False 0x10a6ae520>
alice@16Uiu	  loop.close()
alice@16Uiu	/Users/hwwang/ethereum/trinity/trinity/extensibility/asyncio.py:60: ResourceWarning: unclosed resource <TCPTransport closed=False reading=True 0x7fdc6b716330>
alice@16Uiu	  loop.close()
alice@16Uiu	Process SpawnProcess-5:
alice@16Uiu	Traceback (most recent call last):
alice@16Uiu	  File "/Users/hwwang/.pyenv/versions/3.7.5/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
alice@16Uiu	    self.run()
alice@16Uiu	  File "/Users/hwwang/.pyenv/versions/3.7.5/lib/python3.7/multiprocessing/process.py", line 99, in run
alice@16Uiu	    self._target(*self._args, **self._kwargs)
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/trinity/extensibility/component.py", line 105, in _run_process
alice@16Uiu	    cls.run_process(boot_info)
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/trinity/extensibility/asyncio.py", line 56, in run_process
alice@16Uiu	    loop.run_until_complete(task)
alice@16Uiu	  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/trinity/extensibility/asyncio.py", line 73, in _do_run
alice@16Uiu	    await cls.do_run(boot_info, event_bus)
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/trinity/components/eth2/beacon/component.py", line 173, in do_run
alice@16Uiu	    for service in services
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/p2p/service.py", line 358, in cancellation
alice@16Uiu	    await self.wait(self.events.cancelled.wait())
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/p2p/cancellable.py", line 20, in wait
alice@16Uiu	    return await self.wait_first(awaitable, token=token, timeout=timeout)
alice@16Uiu	  File "/Users/hwwang/ethereum/trinity/p2p/cancellable.py", line 43, in wait_first
alice@16Uiu	    return await token_chain.cancellable_wait(*awaitables, timeout=timeout)
alice@16Uiu	  File "/Users/hwwang/.pyenv/versions/3.7.5/envs/env37a/lib/python3.7/site-packages/cancel_token/token.py", line 176, in cancellable_wait
alice@16Uiu	    "Cancellation requested by {} token".format(self.triggered_token)
alice@16Uiu	cancel_token.exceptions.OperationCancelled: Cancellation requested by BeaconChainSyncer:Node token
alice@16Uiu	Task was destroyed but it is pending!
alice@16Uiu	task: <Task pending coro=<GossipSub.heartbeat() running at /Users/hwwang/.pyenv/versions/3.7.5/envs/env37a/lib/python3.7/site-packages/libp2p/pubsub/gossipsub.py:403> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10a719e50>()]>>
alice@16Uiu	    INFO  2019-12-15 23:47:41,192  AsyncioIsolatedComponent  Process 73624 has already terminated
alice@16Uiu	    INFO  2019-12-15 23:47:41,192  AsyncioIsolatedComponent  Process 73624 has already terminated
alice@16Uiu	    INFO  2019-12-15 23:47:41,192  AsyncioIsolatedComponent  Process 73624 has already terminated
```

[Full log is here](https://gist.github.com/hwwhww/420aa31417944ade68cbc729728155a9)

It seems something else is going wrong after #1379. Both nodes crashed.

### How can it be fixed

Fill this in if you know how to fix it.

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

Fix #1398

### How was it fixed?

- Let syncer run forever
- In `select_sync_peer`, compare peer's head with our head, instead of our finalized head.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: tests/libp2p/bcc/test_syncing.py [hunk 1]
```diff
@@ -23,13 +23,16 @@
...
             chain_db__db=alice.chain.chaindb.db, peer_pool=alice.handshaked_peers
         )
-        asyncio.ensure_future(alice_syncer.run())
 
-        def finalizer():
-            event_loop.run_until_complete(alice_syncer.cancel())
+        timeout = asyncio.Event()
+        async def run_until_timeout():
+            try:
+                await asyncio.wait_for(alice_syncer.run(), timeout=6)
```
Comment (ChihChengLiang) [inline tests/libp2p/bcc/test_syncing.py]: Sorry, I'm writing the magic timeout here that might cause flaky tests. The possible alternatives are 

1. Adding a sync finished event in BeaconChainSyncer just for testing purposes.
2. Emitting a message in event_bus in BeaconChainSyncer just for testing purposes.

Not sure if we have a preference here?

cc @pipermerriam @cburgdorf

Comment (cburgdorf) [reply tests/libp2p/bcc/test_syncing.py]: I'm not necessarily opposed to the timeout itself but I think the fixed value of `6` at this level might cause problems. If I understand correctly then every test that is bootstrapped with `get_sync_setup` will depend on finishing within 6 seconds when in reality they might differ depending on the specific test we are making.

What if instead we change `get_sync_setup` to `get_sync_setup(request, event_loop, event_bus, alice_branch, bob_branch, sync_timeout=6):` to have that tweakable per test?

Also, I think I don't understand the construct with the `asyncio.Event` here. It looks like we want yield `alice` and `bob` after the timeout happened and then do our assertions. If that's the case, then wouldn't this be equivalent?

```python
@asynccontextmanager
async def get_sync_setup(request, event_loop, event_bus, alice_branch, bob_branch, sync_timeout=6):
    alice_chaindb = AsyncBeaconChainDBFactory(blocks=alice_branch)
    bob_chaindb = AsyncBeaconChainDBFactory(blocks=bob_branch)
    peer_pair = ConnectionPairFactory(
        alice_chaindb=alice_chaindb, bob_chaindb=bob_chaindb
    )
    async with peer_pair as (alice, bob):

        alice_syncer = BeaconChainSyncerFactory(
            chain_db__db=alice.chain.chaindb.db, peer_pool=alice.handshaked_peers
        )

        try:
            await asyncio.wait_for(alice_syncer.run(), timeout=sync_timeout)
        except asyncio.TimeoutError:
            # After sync is cancelled, return to the caller to let them do assertions about the state
            pass
        finally:
            yield alice, bob

```

Comment (ChihChengLiang) [reply tests/libp2p/bcc/test_syncing.py]: Oh yes, your proposed customized timeout and the `asyncio.Event` free solution looks nicer!

File: trinity/sync/beacon/chain.py [hunk 2]
```diff
@@ -66,55 +65,32 @@ def __init__(self,
...
-        new_head = await self.chain_db.coro_get_canonical_head(BeaconBlock)
-        self.logger.info(f"Sync with {self.sync_peer} finished, new head: {new_head}")
+                # wait some time and try again
+                await asyncio.sleep(PEER_SELECTION_RETRY_INTERVAL)
+                continue
+
+            # sync peer selected successfully
+            await self.wait(self.sync())
+            new_head = await self.chain_db.coro_get_canonical_head(BeaconBlock)
+            self.logger.info(f"Sync with {self.sync_peer} finished, new head: {new_head}")
```
Comment (hwwhww) [inline trinity/sync/beacon/chain.py]: Should not use f-string for logging.
```suggestion
            self.logger.info(
                "Sync with %s finished, new head: %s",
                self.sync_peer,
                new_head,
            )
```

File: trinity/sync/beacon/chain.py [hunk 3]
```diff
@@ -66,55 +65,32 @@ def __init__(self,
...
-
-        await self.wait(self.sync())
-
-        new_head = await self.chain_db.coro_get_canonical_head(BeaconBlock)
-        self.logger.info(f"Sync with {self.sync_peer} finished, new head: {new_head}")
+                # wait some time and try again
+                await asyncio.sleep(PEER_SELECTION_RETRY_INTERVAL)
+                continue
+
+            # sync peer selected successfully
```
Comment (hwwhww) [inline trinity/sync/beacon/chain.py]: Perhaps use `try-except-else` and move the following lines to `else: ...`.

File: trinity/sync/beacon/chain.py [hunk 4]
```diff
@@ -66,55 +65,32 @@ def __init__(self,
...
 
-        try:
-            finalized_head = await self.chain_db.coro_get_finalized_head(BeaconBlock)
-        # TODO(ralexstokes) look at better way to handle once we have fork choice in place
-        except FinalizedHeadNotFound:
-            return best_peer
-
-        if best_peer.head_slot <= finalized_head.slot:
+        if best_peer.head_slot <= head.slot:
             raise ValidationError("No peer that is ahead of us")
```
Comment (hwwhww) [inline trinity/sync/beacon/chain.py]: nitpick: IMO it doesn't look like a "`ValidationError`". Maybe add a new exception like `LeadingPeerNotFound`

File: trinity/sync/beacon/chain.py [hunk 5]
```diff
@@ -66,55 +65,32 @@ def __init__(self,
...
     async def select_sync_peer(self) -> Peer:
         if len(self.peer_pool) == 0:
             raise ValidationError("Not connected to anyone")
 
         best_peer = self.peer_pool.get_best_head_slot_peer()
+        head = await self.chain_db.coro_get_canonical_head(BeaconBlock)
 
-        try:
-            finalized_head = await self.chain_db.coro_get_finalized_head(BeaconBlock)
-        # TODO(ralexstokes) look at better way to handle once we have fork choice in place
```
Comment (hwwhww) [inline trinity/sync/beacon/chain.py]: I guess it was partially from https://github.com/ethereum/eth2.0-specs/blob/v09x/specs/networking/p2p-interface.md#status

> Once the handshake completes, the client with the lower finalized_epoch or head_slot (if the clients have equal finalized_epochs) SHOULD request beacon blocks from its counterparty via the BeaconBlocksByRange request.

Should we keep `best_peer.head_slot <= finalized_head.slot`?

/cc @ralexstokes

Comment (ChihChengLiang) [reply trinity/sync/beacon/chain.py]: Using `finalized_head.slot` here will cause the syncer to request blocks from the peer even if you are already synced (`head.slot == best_peer.head_slot`). 

```diff
-        if best_peer.head_slot <= finalized_head.slot:
+        if best_peer.head_slot <= head.slot:
```

We do need to refine the syncer more. Here's what I feel the syncer might go wrong and we should do it soon in the subsequent PR.

- We should request the peer's status once for a while to have the updated knowledge of the peer's finalized root and epoch.
- We should have a way to trigger sync immediately after the handshake.

Comment (hwwhww) [reply trinity/sync/beacon/chain.py]: > subsequent PR
Can you open an issue for this?
This PR looks good to me if it's going to be fixed soon.

Comment (ChihChengLiang) [reply trinity/sync/beacon/chain.py]: Opened #1401 for this

**Pull request merged**